keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,305 @@
+
+## Proposal
+
+* Track volume limits for CSI driver in `CSINode` objects instead of `Node`. Limit in `CSINode` is used in-stead of limit coming from `Node` object whenever available for same in-tree volume type.
+  * To get rid of prefix + SHA for `ResourceName` of CSI volumes.
+  * So in-tree volume plugin can share limits with CSI driver that uses the same storage backend.
+
+* Kubelet will create `CSINode` instance during initial CSI Driver registration.
+  * Limits of each CSI volume plugin will be added to `CSINode.spec.drivers[xyz].allocatable`.
+    * All in-tree volume plugins will keep reporting their limits via `Node` object but if a CSI driver is also installed for same driver then limit coming from `CSINode` object will be used.
+
+    * If a CSI driver is registered for an in-tree volume plugin and it reports a different volume limit than in-tree volume plugin, the limit reported by CSI driver is used and kubelet logs a warning.
+  * User may NOT change `CSINode.spec.drivers[xyz].allocatable` to override volume plugin / CSI driver values, e.g. to "reserve" some attachment to the operating system. Kubelet will periodically reconcile `CSINode` and overwrite the value.
+    * Especially, `kubelet --kube-reserved` or `--system-reserved` cannot be used to "reserve" volumes for kubelet or the OS. It is not possible with existing kubelet and this KEP does not change it. We expect that CSI drivers will have configuration options / cmdline arguments to reserve some volumes and they will report their limit already reduced by that reserved amount.
+
+  * Kubelet will continue filling `Node.status.allocatable` and `Node.status.capacity` for both in-tree and CSI volumes during deprecation period.
+    * After deprecation period For CSI volumes limits coming from `Node.status.allocatable` and `Node.status.capacity` will be completely ignored.
+    * For in-tree drivers (i.e AWS, Azure, GCE and Cinder) existing predicates will keep working until CSI migration is enabled by default for each driver. Once CSI migration is enabled by default for a volume plugin, after 2 releases the predicate for that volume plugin will be removed from scheduler. For example: If `CSIMigrationGCE` is enabled by default in 1.15 then in 1.17 - scheduler will completly remove `MaxGCEPDVolumeCountPred` predicate.
+    * If `CSINode.spec.drivers[xyz]` (or whole `CSINode`) is missing, scheduler falls back to `Node.status.allocatable` for volume limits. This solves version skew between old kubelet (using `Node.status`) and new scheduler.
+    * The in-tree drivers for which migration has been enabled by default, their in-tree predicate (such as `MaxGCEPDVolumeCountPred`) will be updated to check both `CSINode.Spec.Drivers` and `CSINode` annotation to verify if migration is enabled. If for a in-tree driver migration is enabled but `CSINode.Spec.Drivers` does not have an entry then scheduling decision will be offloaded to `MaxCSIVolumeCountPred` predicate.
```

Comment:
just enabled, not necessarily by default

Reply:
I'm unclear what offloading to `MaxCSIVolumeCountPred` means. Isn't that the predicate we're designing here?

Maybe it's simpler to say that we will not schedule any pods to a node if migration is enabled but the driver is not installed.

Reply:
ack. fixed.